src/core/models.py

```diff
@@ -0,0 +1,195 @@
+import logging
+from typing import Literal, MutableMapping
+
+from ops.model import Application, Relation, Unit
+from src.literals import CHARM_USERS
+
+logger = logging.getLogger(__name__)
+
+# TODO: move to some better literals place
+SUBSTRATE = Literal["vm", "k8s"]
+CLIENT_PORT = 2181
+SECURE_CLIENT_PORT = 2182
+SERVER_PORT = 2888
+ELECTION_PORT = 3888
+
+class StateBase:
+    def __init__(self, relation: Relation | None, component: Unit | Application, substrate: SUBSTRATE):
+        self.relation = relation
+        self.component = component
+        self.substrate = substrate
```

Comment:
`substrate` is only used on `ZKServer` for the host, so it can be removed from `StateBase` and be added only to `ZKServer`

Reply:
The intent is that we use `StateBase` as a common child class across DP charms. It's also likely that `substrate` will be needed elsewhere, but only `host` right now for this first PR.

I'll be OK with removing it when we get further on, but for now I'm wondering if it'll be needed elsewhere.

Reply:
Tbh, being a literal, you can probably not even pass it around and just read it from somwhere.

Reply:
Honestly you might be right there, and that might be better. I'll hold off on that change for now but add a `FIXME` note.